crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -25,26 +78,99 @@ use ruff_db::system::{System, SystemPath, SystemPathBuf};
 /// will be `/opt/homebrew`, and `site-packages` will be at
 /// `/opt/homebrew/lib/python3.X/site-packages`.
 ///
-/// This routine does not verify that `sys_prefix_path` points
-/// to an existing directory on disk; it is assumed that this has already
-/// been checked.
-///
 /// [`sys.prefix`]: https://docs.python.org/3/library/sys.html#sys.prefix
+#[derive(Debug, PartialEq, Eq)]
+struct SysPrefixPath(SystemPathBuf);
+
+impl SysPrefixPath {
+    fn new(unvalidated_path: impl AsRef<SystemPath>, system: &dyn System) -> Option<Self> {
+        let unvalidated_path = unvalidated_path.as_ref();
+        // It's important to resolve symlinks here rather than simply making the path absolute,
+        // since system Python installations often only put symlinks in the "expected"
+        // locations for `home` and `site-packages`
+        let canonicalized = system.canonicalize_path(unvalidated_path).ok()?;
+        system
+            .is_directory(&canonicalized)
+            .then_some(Self(canonicalized))
+    }
+
+    fn from_executable_home_path(path: &PythonHomePath, system: &dyn System) -> Option<Self> {
+        let candidate = path.parent()?;
+        system
+            .is_directory(candidate)
+            .then(|| Self(candidate.to_path_buf()))
+    }
+}
+
+impl Deref for SysPrefixPath {
+    type Target = SystemPath;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl fmt::Display for SysPrefixPath {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "`sys.prefix` path {}", self.0)
+    }
+}
+
+/// E.g. `12` for Python 3.12.4
+///
+/// Note: this may not represent a Python version that we actually support!
+/// In principle, it's perfectly fine for a user to e.g. pass us the path
+/// to a venv constructed using Python 3.14 for resolving third-party types,
+/// even if we don't yet support Python 3.14, as long as they also pass
+/// e.g. `--target-version=3.12`
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+struct PythonMinorVersion(u8);
+
+impl PythonMinorVersion {
+    fn from_version_string(version_string: &str) -> Option<Self> {
+        version_string.parse().ok().map(Self)
+    }
+
+    const fn free_threaded_build_available(self) -> bool {
+        self.0 >= 13
+    }
+}
+
+impl Deref for PythonMinorVersion {
+    type Target = u8;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl fmt::Display for PythonMinorVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl PartialEq<u8> for PythonMinorVersion {
+    fn eq(&self, other: &u8) -> bool {
+        &self.0 == other
+    }
+}
+
+/// Attempt to retrieve the `site-packages` directory
+/// associated with a given Python installation.
 fn site_packages_dir_from_sys_prefix(
-    sys_prefix_path: &SystemPath,
+    sys_prefix_path: &SysPrefixPath,
+    venv_minor_version: Option<PythonMinorVersion>,
```

Comment:
Not to keep beating the same drum, but this function is not specific to being used from a venv, and there's no need to name the Python minor version this way, it's just a Python minor version no matter where we get it from.
```suggestion
    python_minor_version: Option<PythonMinorVersion>,
```